### ISSUE TITLE
Refactor|Remove Parameter can leave broken code

### ISSUE THREAD

**Issue by daFreeMan** [rubberduck-vba/Rubberduck#4234]
Given some code:

```vba
Private Sub Foo (ByVal Bar as String, ByVal Baz as Long)
  If Baz > 3 Then
    MsgBox "Baz is big!"
  End If
End Sub
```

`Refactor|Remove Parameter` to remove `Baz` leaves me with:

```vba
Public Sub Foo(ByVal Bar As String)
  If Baz > 3 Then
    MsgBox "Baz is big"
  End If
End Sub
```

Which won't compile (with the assumed `Option Explicit`), and won't run. 

The fix could be difficult... 
**A)** Delete all references to `Baz` in the code. This _could_ leave broken code, and is likely to leave some sort of subtle bug.
**B)** Insert `Dim Baz as Long` (type matching that of the deleted parameter) which could leave broken code because `Baz` is not initialized.
**C)** Insert `TODO: Figure out how to handle the removed parameter "Baz as Long"`. Note the explicit reminder of the data type.
**D)** Pop up a dialog box warning that code is now possibly broken due to removal of parameter `Baz` and leave the programmer to deal with it after clicking `OK`.

Either write them all and give the user the choice of how (s)he wants to fix it, or just go with option **C)** as the most useful and longest lasting (option **D)** goes away with the click of a button and can be quickly forgotten)

Of course, if the `Baz` reference count = 0 in the module, none of the above matter and a simple removal of the parameter is all that's needed

**Comment by retailcoder** [comment]
I think the only reasonable "fix" is D. Removing `Baz` in `If Baz > 3 Then` would also break the code, and removing the entire block would be even more damaging.

We should refuse to perform the refactoring until reference count is 0 - possibly pop up an *all references* "search results" tab for it to assist.

**Comment by bclothier** [comment]
I don't think the option D goes far enough - as @retailcoder mentions, we should refuse to remove any parameter that is in use. I think it's best for RD to never refactor anything into a broken code, not even with a `FIXME` comment.

**Comment by daFreeMan** [comment]
@retailcoder - I'd say this is more of a bug than an enhancement. You'll never guess how I came across this one... To be fair, though, the refactoring code does work, it just doesn't leave functional code behind, so I guess it could be considered an enhancement.

I do agree with @bclothier - refusing to remove the parameter while the variable is in use does seem to be a reasonable course of action.

**Comment by retailcoder** [comment]
So the fix is to simply change the command's `CanExecute` logic to account for the `References` of the target.

**Comment by daFreeMan** [comment]
well, maybe... I believe `CanExecute = False` simply disables the command in the menu, right? That leaves no indication of why I can't do it.

IMHO, it would be nicer to have open the dialog and allow me to chose which parameter I want to delete then display a warning that there are references to the chosen parameter, leaving the `OK` button disabled.

Of course, if that's what you meant, then never mind...

**Comment by retailcoder** [comment]
hmm that's right, I forgot the refactoring worked at procedure level and popped a dialog to pick the parameter(s) to remove.

In that case the best UX would probably be to have some :warning: icon for each parameter that's actually used in the body of the procedure, with a tooltip explaining as much. Then if the dialog is okayed anyway, request user confirmation before proceeding, warning that the code will require adjustments to work again.

**Comment by comintern** [comment]
I don't think a simple reference count will determine if it is "safe" to remove a parameter. Wouldn't we also need to determine if it is being used by _calling_ code? Removing a parameter breaks this at the call site:

```
Private Sub Caller()
    Foo "Baz", 10
End Sub

Private Sub Foo(ByVal Bar As String, ByVal Baz As Long)
  If Baz > 3 Then
    MsgBox "Baz is big!"
  End If
End Sub
```
I think that if the parameter is not `Optional`, this refactoring should give a warning if the `procedure` is referenced also.

**Comment by daFreeMan** [comment]
@comintern - the refactor already removes the parameter from all calling sites, so that part is covered.

@retailcoder - I think I can stand behind that idea.

**Comment by comintern** [comment]
Hmmm... I'm not sure if that precludes giving a warning, especially in that the invocation of the function could have side effects:

```
Private Sub Caller()
    Foo "Baz", ReturnSomeValueWithSideEffects()
End Sub
```

**Comment by retailcoder** [comment]
@comintern good point, assuming you're referring to the removal of the 2nd parameter. Although.... if your functions have side-effects, you've been asking for trouble in the first place :wink:

**Comment by comintern** [comment]
@retailcoder That's fair, but you also have to consider non-user functions. This is slightly contrived (best example I could think of on the spot), but what about something more like:

```
Private Sub Caller()
    Foo "Baz", Application.Wait(Now + TimeValue("0:00:10"))
End Sub

Public Sub Foo(message As String, waited As Boolean)
    If waited Then
        MsgBox message
    End If
End Sub
```
"Warning" may be a little strong, but it would be helpful to have something like a count of the calling sites and an option to view the procedure references.
